nifi-nar-bundles/nifi-standard-services/nifi-dbcp-service-bundle/nifi-dbcp-service/src/main/java/org/apache/nifi/record/sink/db/DatabaseRecordSink.java

```diff
@@ -0,0 +1,546 @@
+
+                try {
+                    ps.setQueryTimeout(queryTimeout); // timeout in seconds
+                } catch (SQLException se) {
+                    // If the driver doesn't support query timeout, then assume it is "infinite". Allow a timeout of zero only
+                    if (queryTimeout > 0) {
+                        throw se;
+                    }
+                }
+
+                Record currentRecord;
+                List<Integer> fieldIndexes = sqlHolder.getFieldIndexes();
+
+                while ((currentRecord = recordSet.next()) != null) {
+                    Object[] values = currentRecord.getValues();
+                    List<DataType> dataTypes = currentRecord.getSchema().getDataTypes();
+                    if (values != null) {
+                        if (fieldIndexes != null) {
+                            for (int i = 0; i < fieldIndexes.size(); i++) {
+                                final int currentFieldIndex = fieldIndexes.get(i);
+                                final Object currentValue = values[currentFieldIndex];
+                                final DataType dataType = dataTypes.get(currentFieldIndex);
+                                final int sqlType = DataTypeUtils.getSQLTypeValue(dataType);
+
+                                ps.setObject(i + 1, currentValue, sqlType);
+                            }
+                        } else {
+                            // If there's no index map, assume all values are included and set them in order
+                            for (int i = 0; i < values.length; i++) {
+                                final Object currentValue = values[i];
+                                final DataType dataType = dataTypes.get(i);
+                                final int sqlType = DataTypeUtils.getSQLTypeValue(dataType);
+                                ps.setObject(i + 1, currentValue, sqlType);
+                            }
+                        }
+                        ps.addBatch();
+                    }
+                }
+                ps.executeBatch();
+            }
+
+        } catch (IOException ioe) {
+            throw ioe;
+        } catch (Exception e) {
+            throw new IOException("Failed to write metrics using record writer: " + e.getMessage(), e);
+        } finally {
+            if (connection != null) {
+                if (originalAutoCommit != null) {
+                    try {
+                        connection.setAutoCommit(originalAutoCommit);
+                    } catch (Exception e) {
+                        getLogger().debug("Error restoring auto-commit", e);
+                    }
+                }
+                try {
+                    connection.close();
+                } catch (Exception e) {
+                    getLogger().debug("Error closing connection", e);
+                }
+            }
+        }
+        return null;
```

Comment:
Probably don't want to return null from here.

Reply:
My bad, forgot to implement that. I've added a WriteResult (when successful) and updated the unit tests to verify the contents